File: frontend/src/types/performance_document.ts
/**
 * Типы для документов спектакля.
 */

// Разделы паспорта спектакля
export type DocumentSection = '1.0' | '2.0' | '3.0' | '4.0';

// Категории документов
export type PerformanceDocumentCategory =
  | 'passport' | 'reception_act' | 'fire_protection' | 'welding_acts'
  | 'material_certs' | 'calculations'
  | 'sketches' | 'tech_spec_decor' | 'tech_spec_light' | 'tech_spec_costume'
  | 'tech_spec_props' | 'tech_spec_sound'
  | 'decor_photos' | 'layouts' | 'mount_list' | 'hanging_list'
  | 'mount_instruction' | 'light_partition' | 'sound_partition'
  | 'video_partition' | 'costume_list' | 'makeup_card'
  | 'rider' | 'estimates' | 'drawings'
  | 'other';

// Включение в отчёт
export type ReportInclusion = 'full' | 'partial' | 'excluded';

// Документ спектакля
export interface PerformanceDocument {
  id: number;
  performance_id: number;
  file_path: string;
  file_name: string;
  file_size: number;
  mime_type: string;
  section: DocumentSection;
  category: PerformanceDocumentCategory;
  subcategory?: string;
  display_name: string;
  description?: string;
  sort_order: number;
  report_inclusion: ReportInclusion;
  version: number;
  is_current: boolean;
  uploaded_by_id?: number;
  uploaded_at: string;
  download_url?: string;
}

// Элемент списка документов
export interface PerformanceDocumentListItem {
  id: number;
  file_name: string;
  file_size: number;
  mime_type: string;
  section: DocumentSection;
  category: PerformanceDocumentCategory;
  display_name: string;
  uploaded_at: string;
  download_url?: string;
}

// Категория в дереве
export interface DocumentTreeCategory {
  category: PerformanceDocumentCategory;
  category_name: string;
  documents: PerformanceDocumentListItem[];
  count: number;
}

// Раздел в дереве
export interface DocumentTreeSection {
  section: DocumentSection;
  section_name: string;
  categories: DocumentTreeCategory[];
  total_count: number;
}

// Дерево документов
export interface PerformanceDocumentsTree {
  performance_id: number;
  sections: DocumentTreeSection[];
  total_documents: number;
}

// Обновление документа
export interface PerformanceDocumentUpdate {
  display_name?: string;
  description?: string;
  section?: DocumentSection;
  category?: PerformanceDocumentCategory;
  subcategory?: string;
  report_inclusion?: ReportInclusion;
  sort_order?: number;
}

// Готовность раздела паспорта
export interface PassportSectionReadiness {
  section: string;
  section_name: string;
  progress: number;
  status: 'EMPTY' | 'IN_PROGRESS' | 'COMPLETE';
  filled_categories: number;
  total_categories: number;
}

// Готовность паспорта спектакля
export interface PassportReadiness {
  overall_progress: number;
  sections: PassportSectionReadiness[];
}

// Детализированная готовность раздела
export interface SectionDetailedReadiness {
  section: string;
  section_name: string;
  progress: number;
  filled_categories: number;
  total_categories: number;
  categories: {
    category: PerformanceDocumentCategory;
    category_name: string;
    required: boolean;
    filled: boolean;
    documents_count: number;
  }[];
}

// Названия разделов
export const SECTION_NAMES: Record<DocumentSection, string> = {
  '1.0': '1.0 Общая часть',
  '2.0': '2.0 Производство',
  '3.0': '3.0 Эксплуатация',
  '4.0': '4.0 Приложение',
};

// Названия категорий
export const CATEGORY_NAMES: Record<PerformanceDocumentCategory, string> = {
  passport: 'Паспорт спектакля',
  reception_act: 'Акт приёмки декораций',
  fire_protection: 'Огнезащитная обработка',
  welding_acts: 'Акты сварочных работ',
  material_certs: 'Сертификаты материалов',
  calculations: 'Расчёты конструкций',
  sketches: 'Эскизы',
  tech_spec_decor: 'ТЗ декорация',
  tech_spec_light: 'ТЗ свет',
  tech_spec_costume: 'ТЗ костюм',
  tech_spec_props: 'ТЗ реквизит',
  tech_spec_sound: 'ТЗ звук',
  decor_photos: 'Фото декораций',
  layouts: 'Планировки',
  mount_list: 'Монтировочная опись',
  hanging_list: 'Ведомость развески',
  mount_instruction: 'Инструкция монтажа',
  light_partition: 'Партитура света',
  sound_partition: 'Партитура звука',
  video_partition: 'Партитура видео',
  costume_list: 'Опись костюмов',
  makeup_card: 'Грим-карта',
  rider: 'Райдер',
  estimates: 'Сметы',
  drawings: 'Чертежи',
  other: 'Прочее',
};

// Цвета разделов
export const SECTION_COLORS: Record<DocumentSection, string> = {
  '1.0': '#3B82F6',  // blue
  '2.0': '#8B5CF6',  // purple
  '3.0': '#F59E0B',  // amber
  '4.0': '#10B981',  // emerald
};

// Иконки по MIME типу
export const FILE_TYPE_ICONS: Record<string, string> = {
  'application/pdf': 'document',
  'image/': 'photo',
  'audio/': 'music',
  'video/': 'video',
  'application/vnd.openxmlformats-officedocument.spreadsheetml': 'table',
  'application/vnd.ms-excel': 'table',
  'default': 'paper-clip',
};

// Форматирование размера файла
export function formatFileSize(bytes: number): string {
  if (bytes === 0) return '0 Б';
  const k = 1024;
  const sizes = ['Б', 'КБ', 'МБ', 'ГБ'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + ' ' + sizes[i];
}

// Определение иконки по MIME типу
export function getFileIcon(mimeType: string): string {
  for (const [key, icon] of Object.entries(FILE_TYPE_ICONS)) {
    if (mimeType.startsWith(key) || mimeType === key) {
      return icon;
    }
  }
  return FILE_TYPE_ICONS['default'];
}

File: backend/app/schemas/performance_document.py
"""
Pydantic схемы для документов спектакля.
"""
from datetime import datetime
from pydantic import BaseModel, Field

from app.models.performance_document import (
    DocumentSection,
    PerformanceDocumentCategory,
    ReportInclusion,
)


# =============================================================================
# Request Schemas
# =============================================================================

class PerformanceDocumentCreate(BaseModel):
    """Создание документа (метаданные для загрузки)."""

    section: DocumentSection | None = Field(
        None,
        description="Раздел паспорта (определяется автоматически если не указан)"
    )
    category: PerformanceDocumentCategory | None = Field(
        None,
        description="Категория документа (определяется автоматически если не указана)"
    )
    display_name: str | None = Field(
        None,
        description="Отображаемое имя (по умолчанию — имя файла)"
    )
    description: str | None = None
    report_inclusion: ReportInclusion = Field(
        default=ReportInclusion.FULL,
        description="Включение в отчёт"
    )


class PerformanceDocumentUpdate(BaseModel):
    """Обновление метаданных документа."""

    display_name: str | None = None
    description: str | None = None
    section: DocumentSection | None = None
    category: PerformanceDocumentCategory | None = None
    subcategory: str | None = None
    report_inclusion: ReportInclusion | None = None
    sort_order: int | None = None


# =============================================================================
# Response Schemas
# =============================================================================

class PerformanceDocumentResponse(BaseModel):
    """Ответ с документом спектакля."""

    id: int
    performance_id: int
    file_path: str
    file_name: str
    file_size: int
    mime_type: str
    section: DocumentSection
    category: PerformanceDocumentCategory
    subcategory: str | None = None
    display_name: str
    description: str | None = None
    sort_order: int
    report_inclusion: ReportInclusion
    version: int
    is_current: bool
    uploaded_by_id: int | None = None
    uploaded_at: datetime
    download_url: str | None = None

    model_config = {"from_attributes": True}


class PerformanceDocumentListItem(BaseModel):
    """Элемент списка документов."""

    id: int
    file_name: str
    file_size: int
    mime_type: str
    section: DocumentSection
    category: PerformanceDocumentCategory
    display_name: str
    uploaded_at: datetime
    download_url: str | None = None

    model_config = {"from_attributes": True}


class DocumentTreeCategory(BaseModel):
    """Категория в дереве документов."""

    category: PerformanceDocumentCategory
    category_name: str
    documents: list[PerformanceDocumentListItem]
    count: int


class DocumentTreeSection(BaseModel):
    """Раздел в дереве документов."""

    section: DocumentSection
    section_name: str
    categories: list[DocumentTreeCategory]
    total_count: int


class PerformanceDocumentsTree(BaseModel):
    """Дерево документов спектакля."""

    performance_id: int
    sections: list[DocumentTreeSection]
    total_documents: int


class BulkUploadResult(BaseModel):
    """Результат массовой загрузки."""

    uploaded: list[PerformanceDocumentResponse]
    failed: list[dict]  # {filename, error}
    total_uploaded: int
    total_failed: int


# =============================================================================
# Passport Readiness Schemas
# =============================================================================

class PassportSectionReadiness(BaseModel):
    """Готовность раздела паспорта."""

    section: str  # "1.0", "2.0", "3.0", "4.0"
    section_name: str
    progress: int  # 0-100
    status: str  # "EMPTY", "IN_PROGRESS", "COMPLETE"
    filled_categories: int
    total_categories: int


class PassportReadinessResponse(BaseModel):
    """Общая готовность паспорта спектакля."""

    overall_progress: int  # 0-100
    sections: list[PassportSectionReadiness]


class SectionDetailedReadiness(BaseModel):
    """Детализированная готовность раздела."""

    section: str
    section_name: str
    progress: int
    filled_categories: int
    total_categories: int
    categories: list[dict]  # {category, category_name, required, filled, documents_count}


# =============================================================================
# Section/Category Names
# =============================================================================

SECTION_NAMES = {
    DocumentSection.GENERAL: "1.0 Общая часть",
    DocumentSection.PRODUCTION: "2.0 Производство",
    DocumentSection.OPERATION: "3.0 Эксплуатация",
    DocumentSection.APPENDIX: "4.0 Приложение",
}

CATEGORY_NAMES = {
    PerformanceDocumentCategory.PASSPORT: "Паспорт спектакля",
    PerformanceDocumentCategory.RECEPTION_ACT: "Акт приёмки декораций",
    PerformanceDocumentCategory.FIRE_PROTECTION: "Огнезащитная обработка",
    PerformanceDocumentCategory.WELDING_ACTS: "Акты сварочных работ",
    PerformanceDocumentCategory.MATERIAL_CERTS: "Сертификаты материалов",
    PerformanceDocumentCategory.CALCULATIONS: "Расчёты конструкций",
    PerformanceDocumentCategory.SKETCHES: "Эскизы",
    PerformanceDocumentCategory.TECH_SPEC_DECOR: "ТЗ декорация",
    PerformanceDocumentCategory.TECH_SPEC_LIGHT: "ТЗ свет",
    PerformanceDocumentCategory.TECH_SPEC_COSTUME: "ТЗ костюм",
    PerformanceDocumentCategory.TECH_SPEC_PROPS: "ТЗ реквизит",
    PerformanceDocumentCategory.TECH_SPEC_SOUND: "ТЗ звук",
    PerformanceDocumentCategory.DECOR_PHOTOS: "Фото декораций",
    PerformanceDocumentCategory.LAYOUTS: "Планировки",
    PerformanceDocumentCategory.MOUNT_LIST: "Монтировочная опись",
    PerformanceDocumentCategory.HANGING_LIST: "Ведомость развески",
    PerformanceDocumentCategory.MOUNT_INSTRUCTION: "Инструкция монтажа",
    PerformanceDocumentCategory.LIGHT_PARTITION: "Партитура света",
    PerformanceDocumentCategory.SOUND_PARTITION: "Партитура звука",
    PerformanceDocumentCategory.VIDEO_PARTITION: "Партитура видео",
    PerformanceDocumentCategory.COSTUME_LIST: "Опись костюмов",
    PerformanceDocumentCategory.MAKEUP_CARD: "Грим-карта",
    PerformanceDocumentCategory.RIDER: "Райдер",
    PerformanceDocumentCategory.ESTIMATES: "Сметы",
    PerformanceDocumentCategory.DRAWINGS: "Чертежи",
    PerformanceDocumentCategory.OTHER: "Прочее",
}
